### PR TITLE
Fall back to most-recent interop run when SHA is latest

### PR DESCRIPTION
**Pull request by lukebjerring**
## Description
Works towards preventing https://github.com/web-platform-tests/wpt.fyi/issues/514

## Review Information
To test, you'll need to manually ensure that the latest `complete` run does not have interop data, by deleting it if it does.

- Visit the [interop](https://interop-fallback-dot-wptdashboard-staging.appspot.com/interop) tab
  - Observe that you can see some interop scores for a complete run, and what the SHA is
- Visit the corresponding [results](https://interop-fallback-dot-wptdashboard-staging.appspot.com/results) (Test Results) tab
  - Observe that the SHA is different and the runs more recent

## Changes
To achieve this behaviour, `GetCompleteSHAs` has been rejigged to return not only the SHAs in chronological order, but a map of SHA to TestRun keys.

`LoadTestRuns` is now a wrapper for calls to both `LoadTestRunKeys` and `LoadTestRunsForKeys`. This ensures that we can run key-only queries in different contexts efficiently.

Interop has also changed from "find the (one run only) test runs matching the query, then find the single interop data that covers those test runs" to "find up to 128 test runs that match the query, then iterate through the interop runs until we find one where all 4 `TestRunIDs` are present in the set of keys that matched the query.

### CODE REVIEW

File: shared/datastore.go [hunk 1]
```diff
@@ -21,23 +21,20 @@ func LoadTestRun(ctx context.Context, id int64) (*TestRun, error) {
 	return &testRun, nil
 }
 
-// LoadTestRuns loads the TestRun entities for the given parameters.
+// LoadTestRunKeys loads the keys for the TestRun entities for the given parameters.
```
Comment (mdittmer) [inline shared/datastore.go]: Can we get tests for the two new functions in this file?

Comment (mdittmer) [reply shared/datastore.go]: Related note: GitHub's crappy diff is making it hard to do a thorough review of this file. I'm sure once I review your tests for the functions I'll be confident that the code is correct :smiley:

File: shared/util.go [hunk 2]
```diff
@@ -16,6 +16,9 @@ import (
 // ExperimentalLabel is the implicit label present for runs marked 'experimental'.
 const ExperimentalLabel = "experimental"
 
+// LatestSHA is a helper for the 'latest' keyword/special-case.
```
Comment (mdittmer) [inline shared/util.go]: nit: `special case`.

File: util/populate_dev_data.go [hunk 3]
```diff
@@ -229,19 +229,22 @@ func copyProdRuns(ctx context.Context, filters shared.TestRunFilter) {
...
-		var localRunCopies []shared.TestRun
-		localRunCopies, err = shared.LoadTestRuns(ctx, shared.GetDefaultProducts(), filters.Labels, shas, nil, nil, &one)
-		if len(localRunCopies) != len(prodPassRateMetadata.TestRunIDs) {
-			sha := "latest"
+			var shas []string
+			var keys map[string][]*datastore.Key
+			shas, keys, _ = shared.GetCompleteRunSHAs(ctx, shared.GetDefaultProducts(), filters.Labels, nil, nil, &one)
 			if len(shas) > 0 {
 				sha = shas[0]
+				localRunCopies, _ = shared.LoadTestRunsByKeys(ctx, keys[sha])
```
Comment (mdittmer) [inline util/populate_dev_data.go]: Check for and log error?

File: util/populate_dev_data.go [hunk 4]
```diff
@@ -229,19 +229,22 @@ func copyProdRuns(ctx context.Context, filters shared.TestRunFilter) {
...
 		if complete {
-			shas, _ = shared.GetCompleteRunSHAs(ctx, shared.GetDefaultProducts(), filters.Labels, nil, nil, &one)
-		}
-		var localRunCopies []shared.TestRun
-		localRunCopies, err = shared.LoadTestRuns(ctx, shared.GetDefaultProducts(), filters.Labels, shas, nil, nil, &one)
-		if len(localRunCopies) != len(prodPassRateMetadata.TestRunIDs) {
-			sha := "latest"
+			var shas []string
+			var keys map[string][]*datastore.Key
+			shas, keys, _ = shared.GetCompleteRunSHAs(ctx, shared.GetDefaultProducts(), filters.Labels, nil, nil, &one)
```
Comment (mdittmer) [inline util/populate_dev_data.go]: Check for and log error?

File: shared/datastore.go [hunk 5]
```diff
@@ -164,7 +164,8 @@ func VersionPrefix(query *datastore.Query, fieldName, versionPrefix string, desc
 }
 
 // GetCompleteRunSHAs returns an array of the SHA[0:10] for runs that
-// exists for all the given products, ordered by most-recent.
+// exists for all the given products, ordered by most-recent, as well as a map
+// of those SHAs to the TestRun keys for the complete run.
```
Comment (mdittmer) [inline shared/datastore.go]: nit: Discourage this sense of the word "complete".

**Comment by lukebjerring** [comment]
Note that this also fixes #451

File: webdriver/label_test.go [hunk 6]
```diff
@@ -31,11 +31,8 @@ func TestLabelParam(t *testing.T) {
 	} else {
 		// Local static data only have 2 experimental browsers.
 		testLabel(t, wd, app, "/", "experimental", "wpt-results", 2)
-		// TODO(Hexcles): "/interop?label=" is broken on staging (#451),
-		// so we only run this test locally. Once the regression is
-		// fixed, we should run it against staging, too.
-		testLabel(t, wd, app, "/interop", "stable", "wpt-interop", 4)
 	}
+	testLabel(t, wd, app, "/interop", "stable", "wpt-interop", 4)
```
Comment (lukebjerring) [inline webdriver/label_test.go]: Note that this was tested locally against the deployment of this PR, by executing:

    make go_firefox_test FIREFOX_PATH=~/browsers/firefox/firefox FLAGS="-remote_host=interop-fallback-dot-wptdashboard-staging.appspot.com -staging=true -test.run=Label"

File: api/interop.go [hunk 7]
```diff
@@ -18,53 +20,123 @@ import (
...
+				http.Error(w, "No metrics runs found", http.StatusNotFound)
+				return
+			}
+			keysChecker = checkKeysMatchQuery(keys)
+		}
+	}
+
+	// Iterate until we find interop data where its TestRunIDs match the query.
+	var interop metrics.PassRateMetadata
+	it := query.Run(ctx)
```
Comment (Hexcles) [inline api/interop.go]: I'm not seeing any explicit limit on the query. In the worst case, how many iterations will we do?

Comment (lukebjerring) [reply api/interop.go]: I'll cap to 1000. Though, interestingly, this points out a problem with loading really, really old interop data (e.g. for a very specific SHA).

**Comment by lukebjerring** [comment]
I've deployed this as a staging branch in the production environment to better demonstrate its effect.
https://interop-fallback-dot-wptdashboard.appspot.com/interop/?label=stable&aligned=true

File: api/interop.go [hunk 8]
```diff
@@ -18,53 +20,123 @@ import (
...
+	keysFilter := mapset.NewSet()
+	for _, key := range keys {
+		keysFilter.Add(key.IntID())
+	}
+	return func(ids shared.TestRunIDs) bool {
+		all := true
+		for _, id := range ids {
+			if !keysFilter.Contains(id) {
+				all = false
+				break
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply api/interop.go]: Uh, not quite. We're in a nested loop here, so need to consider each of the other sets of `aligned` keys (returning false would return on the first check, ignoring all other aligned sets).

File: api/interop.go [hunk 9]
```diff
@@ -18,53 +20,123 @@ import (
...
+				return
+			} else if len(shaKeys) < 1 {
+				// Bail out early - can't find any complete runs.
+				http.Error(w, "No metrics runs found", http.StatusNotFound)
+				return
+			}
+			keysChecker = checkKeysAreAligned(shaKeys)
+		} else {
+			// Load default browser runs for SHA.
+			// Force any max-count to one; more than one of each product makes no sense for a interop run.
```
[the remark this answers is not in the document]
Comment (lukebjerring) [reply api/interop.go]: Revised.

File: api/interop.go (hunk 8 above)
[the remark this answers is not in the document]
Comment (Hexcles) [reply api/interop.go]: Hmm, I don't see a nested loop here; there's one in `checkKeysAreAligned`. Anyway, it doesn't matter much.

File: api/interop.go (hunk 7 above)
Comment (Hexcles) [reply api/interop.go]: Let's make this 100.

**Comment by Hexcles** [comment]
`sha` by itself works: https://interop-fallback-dot-wptdashboard.appspot.com/interop/?sha=9614def367

**Comment by lukebjerring** [comment]
So, the problem was a lack of support for the legacy format of the data. I've included the fix, but also I've deleted runs with the nested entities in datastore (both prod and staging), which tees us up to remove the legacy struct in a follow-up PR.

**Pull request merged**